nixos/modules/services/monitoring/certspotter.nix

```diff
@@ -0,0 +1,126 @@
+{ config
+, lib
+, pkgs
+, ... }:
+
+let
+  cfg = config.services.certspotter;
+
+  configDir = pkgs.linkFarm "certspotter-config"
+    (lib.toList {
+      name = "watchlist";
+      path = pkgs.writeText "cerspotter-watchlist" (builtins.concatStringsSep "\n" cfg.watchlist);
+    }
+    ++ lib.optional (cfg.emailRecipients != [ ]) {
+      name = "email_recipients";
+      path = pkgs.writeText "cerspotter-email_recipients" (builtins.concatStringsSep "\n" cfg.emailRecipients);
+    }
+    ++ lib.optional (cfg.hooks != [ ]) {
+      name = "hooks.d";
+      path = pkgs.linkFarm "certspotter-hooks" (lib.imap1 (i: path: {
+        inherit path;
+        name = "hook${toString i}";
+      }) cfg.hooks);
+    });
+in
+{
+  options.services.certspotter = {
+    enable = lib.mkEnableOption "Cert Spotter, a Certificate Transparency log monitor";
+
+    sendmailPath = lib.mkOption {
+      type = lib.types.path;
+      description = ''
+        Path to the `sendmail` binary. By default, the local sendmail wrapper is used
+        (see `config.services.mail.sendmailSetuidWrapper`).
+      '';
+      example = lib.literalExpression ''"''${pkgs.system-sendmail}/bin/sendmail"'';
+    };
+
+    watchlist = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Domain names to watch. To monitor a domain with all subdomains, prefix its name with `.` (e.g. `.example.org`).";
+      default = [ ];
+      example = [ ".example.org" "another.example.com" ];
+    };
+
+    emailRecipients = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "A list of email addresses to send certificate updates to.";
+      default = [ ];
+    };
+
+    hooks = lib.mkOption {
+      type = with lib.types; listOf path;
+      description = ''
+        Scripts to run upon the detection of a new certificate. See `man 8 certspotter-script` or [the GitHub page](https://github.com/SSLMate/certspotter/blob/master/man/certspotter-script.md) for more info.
+      '';
+      default = [];
+      example = lib.literalExpression ''
+        [
+          (pkgs.writeShellScript "certspotter-hook" '''
+            echo "Event summary: $SUMMARY."
+          ''')
+        ]
+      '';
+    };
+
+    extraFlags = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Extra command-line arguments to pass to Cert Spotter";
+      example = [ "-start_at_end" ];
+      default = [ ];
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.watchlist != [ ];
+        message = "You must specify at least one domain for Cert Spotter to watch";
```

Comment:
I think it makes sense to describe why this is an assertion. (The only reason I can come up with is that, for people who will ever use only a single-item watchlist, this prevents the mistake where one first starts certspotter without `start_at_end` with an empty watchlist, and then adds the singular watchlist entry once it has caught up. However, that doesn't help people who want more than one entry.)

Reply:
I simply couldn't think of a valid reason to use an empty watchlist

Reply:
An obvious usecase is writing a module that sets up certspotter for a list of domains constructed from some source (or sources). Having to specially handle the case of "the list came up empty" is not helpful.

Reply:
I don't think it's a good idea to enable certspotter if the watchlist is empty, so the special handling has to be *somewhere*. Of course, it's possible to remove `enable` (or default it to true) and only use `watchlist` for checking whether certspotter should run, but that goes against the norm.

Reply:
Why isn't it a good idea? One might want to use certspotter as a CT log consistency checker (i.e. something that alerts if the log does not uphold its guarantee to be append-only).

Reply:
As I wrote above, I "couldn't think of a valid reason to use an empty watchlist". Now that I know there is a valid reason, I'll remove this assertion.